seastar/src/api_safety.rs

```diff
@@ -0,0 +1,82 @@
+#[cxx::bridge(namespace = "seastar")]
+mod ffi {
+    unsafe extern "C++" {
+        include!("seastar/core/reactor.hh");
+
+        /// Checks whether the current thread is within a Seastar runtime.
+        fn engine_is_ready() -> bool;
+    }
+}
+
+pub use ffi::engine_is_ready;
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::AppTemplate;
+    use std::thread;
+
+    #[test]
+    fn test_engine_is_ready_only_for_thread_in_runtime() {
+        assert!(!engine_is_ready());
+
+        thread::spawn(|| {
+            assert!(!engine_is_ready());
+            let _guard = crate::acquire_guard_for_seastar_test();
+            let mut app = AppTemplate::default();
+            let fut = async {
+                assert!(engine_is_ready());
+                Ok(())
+            };
+            app.run_void(std::env::args(), fut);
+            // Currently, Seastar does not perform a cleanup of its
+            // thread locals, which unfortunately means
+            // the following assertion  will pass.
+            assert!(engine_is_ready());
+        })
+        .join()
+        .unwrap();
+
+        assert!(!engine_is_ready());
+    }
+}
+
+/// Intended to be used in a runtime-dependent function.
+/// Panics if called outside of a Seastar runtime.
+#[macro_export]
+macro_rules! assert_runtime_is_running {
+    () => {{
+        fn f() {}
+        fn type_name_of<T>(_: T) -> &'static str {
+            std::any::type_name::<T>()
+        }
+        let name = type_name_of(f);
```

Comment:
I don't understand this code, please explain what is going on here

Reply:
AFAIK most of the code here is a workaround to simulate something like `__FUNCTION__` in C. The type name for `f` function probably includes the name of the outer function and `f()` is later dropped with the `- 3` trick.

It's a bit hacky and I'm not sure it is guaranteed to work in the future. I'm not aware of any alternatives. Perhaps we could drop it? If somebody wants to check the source of the panic, they can set `RUST_BACKTRACE=1` or `RUST_BACKTRACE=full`.

Reply:
There is also `std::backtrace` but its API is not really usable for this purpose yet: https://doc.rust-lang.org/std/backtrace/struct.Backtrace.html

Reply:
If this is a hacky workaround, the it should at least be properly explained in a comment, and it's not

Reply:
Yes, it is precisely as @piodul said. How this works: we create a dummy function and get its type (and the type is qualified with the exact path to the function's declaration), we then cut `"::f` from the type to get the "path" to the parent-function, which in our case will be the one in which the assertion is called.

If you feel like this is unnecessary, and would rather see a more generic message like "invalid use of a runtime-dependent function outside a Seastar runtime", or no message at all (?), please do tell.

Reply:
It's really unfortunate how a "`function!`" macro isn't in stable Rust, but `line!`, `column!`, `module_path!` and `file!` are...

Reply:
I think we don't need that - `RUST_BACKTRACE=1` is enough